src/database/api/Query.ts

```diff
@@ -0,0 +1,515 @@
+      throw new Error('Query.orderByChild: "$priority" is invalid.  Use Query.orderByPriority() instead.');
+    } else if (path === '$value') {
+      throw new Error('Query.orderByChild: "$value" is invalid.  Use Query.orderByValue() instead.');
+    }
+    validatePathString('Query.orderByChild', 1, path, false);
+    this.validateNoPreviousOrderByCall_('Query.orderByChild');
+    const parsedPath = new Path(path);
+    if (parsedPath.isEmpty()) {
+      throw new Error('Query.orderByChild: cannot pass in empty path.  Use Query.orderByValue() instead.');
+    }
+    const index = new PathIndex(parsedPath);
+    const newParams = this.queryParams_.orderBy(index);
+    Query.validateQueryEndpoints_(newParams);
+
+    return new Query(this.repo, this.path, newParams, /*orderByCalled=*/true);
+  }
+
+  /**
+   * Return a new query ordered by the KeyIndex
+   * @return {!Query}
+   */
+  orderByKey(): Query {
+    validateArgCount('Query.orderByKey', 0, 0, arguments.length);
+    this.validateNoPreviousOrderByCall_('Query.orderByKey');
+    const newParams = this.queryParams_.orderBy(KEY_INDEX);
+    Query.validateQueryEndpoints_(newParams);
+    return new Query(this.repo, this.path, newParams, /*orderByCalled=*/true);
+  }
+
+  /**
+   * Return a new query ordered by the PriorityIndex
+   * @return {!Query}
+   */
+  orderByPriority(): Query {
+    validateArgCount('Query.orderByPriority', 0, 0, arguments.length);
+    this.validateNoPreviousOrderByCall_('Query.orderByPriority');
+    const newParams = this.queryParams_.orderBy(PRIORITY_INDEX);
+    Query.validateQueryEndpoints_(newParams);
+    return new Query(this.repo, this.path, newParams, /*orderByCalled=*/true);
+  }
+
+  /**
+   * Return a new query ordered by the ValueIndex
+   * @return {!Query}
+   */
+  orderByValue(): Query {
+    validateArgCount('Query.orderByValue', 0, 0, arguments.length);
+    this.validateNoPreviousOrderByCall_('Query.orderByValue');
+    const newParams = this.queryParams_.orderBy(VALUE_INDEX);
+    Query.validateQueryEndpoints_(newParams);
+    return new Query(this.repo, this.path, newParams, /*orderByCalled=*/true);
+  }
+
+  /**
+   * @param {number|string|boolean|null} value
+   * @param {?string=} name
+   * @return {!Query}
+   */
+  startAt(value: number | string | boolean | null = null, name?: string | null): Query {
+    validateArgCount('Query.startAt', 0, 2, arguments.length);
+    validateFirebaseDataArg('Query.startAt', 1, value, this.path, true);
+    validateKey('Query.startAt', 2, name, true);
+
+    const newParams = this.queryParams_.startAt(value, name);
+    Query.validateLimit_(newParams);
+    Query.validateQueryEndpoints_(newParams);
+    if (this.queryParams_.hasStart()) {
+      throw new Error('Query.startAt: Starting point was already set (by another call to startAt ' +
+        'or equalTo).');
+    }
+
+    // Calling with no params tells us to start at the beginning.
+    if (value == null) {
```

Comment:
There's a parameter initializer on `value` (`= null`) so by the time it reaches this check it will never be undefined.